encryption/x25519/pubKeyEncryptor.go

```diff
@@ -0,0 +1,248 @@
+//  private key generated on the spot. The senderPrivateKey param is used to authenticate the encryption
+//  that normally should happen between two edwards curve identities.
+func (ed *EncryptedData) Encrypt(data []byte, recipientPubKey crypto.PublicKey, senderPrivateKey crypto.PrivateKey) error {
+	suite := ed25519.NewEd25519()
+	ephemeralEdScalar, ephemeralEdPoint := suite.CreateKeyPair()
+
+	recipientPubKeyBytes, err := recipientPubKey.ToByteArray()
+	if err != nil {
+		return err
+	}
+
+	nonce, err := ed.generateEncryptionNonce(data)
+	if err != nil {
+		return err
+	}
+
+	ciphertext, err := ed.createCiphertext(data, ephemeralEdScalar, recipientPubKey, nonce)
+	if err != nil {
+		return err
+	}
+
+	ephemeralEdPointBytes, err := ephemeralEdPoint.MarshalBinary()
+	if err != nil {
+		return err
+	}
+	mac, err := ed.generateMAC(senderPrivateKey, append(ciphertext, ephemeralEdPointBytes...))
+
+	senderPubKey, err := senderPrivateKey.GeneratePublic().ToByteArray()
+	if err != nil {
+		return err
+	}
+
+	ed.Nonce = hex.EncodeToString(nonce)
+	ed.Version = EncryptionVersion
+	ed.Crypto.Cipher = EncryptionCipher
+	ed.Crypto.Ciphertext = hex.EncodeToString(ciphertext)
+	ed.Crypto.MAC = hex.EncodeToString(mac)
+	ed.Identities.EphemeralPubKey = hex.EncodeToString(ephemeralEdPointBytes)
+	ed.Identities.OriginatorPubKey = hex.EncodeToString(senderPubKey)
+	ed.Identities.Recipient = hex.EncodeToString(recipientPubKeyBytes)
+
+	return nil
+}
+
+// Decrypt returns the plain text associated to a ciphertext that was previously encrypted
+//  using the public key of the recipient
+func (ed *EncryptedData) Decrypt(recipientPrivateKey crypto.PrivateKey) ([]byte, error) {
+	var nonce24 [24]byte
+	var pubKey32 [32]byte
+	var secretKey32 [32]byte
+
+	nonce, err := hex.DecodeString(ed.Nonce)
+	if err != nil {
+		return nil, err
+	}
+	encryptedMessage, err := hex.DecodeString(ed.Crypto.Ciphertext)
+	if err != nil {
+		return nil, err
+	}
+	copy(nonce24[:], nonce)
+
+	pubKeyBytes, err := hex.DecodeString(ed.Identities.EphemeralPubKey)
+	if err != nil {
+		return nil, err
+	}
+
+	originatorPubKeyBytes, err := hex.DecodeString(ed.Identities.OriginatorPubKey)
+	if err != nil {
+		return nil, err
+	}
+	macBytes, err := hex.DecodeString(ed.Crypto.MAC)
+	if err != nil {
+		return nil, err
+	}
+
+	// Verification of the auth message
+	message := append(encryptedMessage, pubKeyBytes...)
+	authMessage := sha256.New()
+	_, err = authMessage.Write(message)
+	if err != nil {
+		return nil, err
+	}
+
+	suite := ed25519.NewEd25519()
+	keygen := crypto.NewKeyGenerator(suite)
+	originatorPubKey, err := keygen.PublicKeyFromByteArray(originatorPubKeyBytes)
+	if err != nil {
+		return nil, err
+	}
+	signer := singlesig.Ed25519Signer{}
+	err = signer.Verify(originatorPubKey, authMessage.Sum(nil), macBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	var recipientX25519 x25519Point
```

Comment:
can you extract the decryption in a private decrypt function and call it here?

Reply:
done